### PR TITLE
Drop support for Mapping types as Selection options

### LINKED ISSUE #1958: Drop support for mutable types as dropdown options
I know this has been reviewed before, but I humbly request it be reviewed again.. :-)

I don't see any upside to supporting dictionaries as an options list (you can just use `widget.options = dict.items()` if you want the same behaviour, which more clearly separates ownership semantics), and it is a source of many pitfalls.  I suspect that many people use dictionaries simply because it's more prominently featured in the docs than the list-of-tuples approach.

I realize this is a backwards-incompatible change, but I think moving forward it would reduce community support burden (and frustrated user time).  Certainly this could be done in a phased manner with `DeprecationWarning` (making sure users can actually see it, of course).

### PR DESCRIPTION
**Pull request by jtpio**
Fixes #1958.
Continuation of #2130. 

### Breaking change

- This removes support for `Mapping` types when constructing a new `Dropdown`.

### CODE REVIEW

**Comment by pbugnion** [comment]
Thanks!

Currently, the `options` traitlet has type `Any()`. I wonder if we can change it to a `traitlets.List` or some such?

I think we can probably cut more code from `_Selection` in general.

**Comment by jtpio** [comment]
That's a good point, let's specify the type even more.

**Comment by jtpio** [comment]
> Currently, the `options` traitlet has type `Any()`. I wonder if we can change it to a `traitlets.List` or some such?

`traitlets.List` might be slightly too specific, since `options` should also support other iterables such as `Tuple`.

**Comment by pbugnion** [comment]
There is no reasonable traitlet supertype that includes general Python iterables, as far as I can see. 

For instance, the following currently works:

```python
import numpy as np
options = np.array(['hello', 'world'])
widgets.Dropdown(options=options)
```

As far as I can tell, no refinement of `Any` would let us keep this behaviour, so I vote in favour of keeping `Any`.

File: ipywidgets/widgets/widget_selection.py [hunk 1]
```diff
@@ -133,9 +127,9 @@ class _Selection(DescriptionWidget, ValueWidget, CoreWidget):
     """Base class for Selection widgets
 
     ``options`` can be specified as a list of values, list of (label, value)
-    tuples, or a dict of {label: value}. The labels are the strings that will be
-    displayed in the UI, representing the actual Python choices, and should be
-    unique. If labels are not specified, they are generated from the values.
+    tuples. The labels are the strings that will be displayed in the UI,
+    representing the actual Python choices, and should be unique.
+    If labels are not specified, they are generated from the values.
```
Comment (pbugnion) [inline ipywidgets/widgets/widget_selection.py]: I think this is missing an `or`:

```rst
``options`` can be specified as a list of values or a list of (label, value) tuples.
```

**Comment by pbugnion** [comment]
I think some of the entries in `doc_snippets` also need updating (e.g. multiple_selection_params).

File: ipywidgets/widgets/widget_selection.py (hunk 1 above)
Comment (jtpio) [reply ipywidgets/widgets/widget_selection.py]: Thanks for catching that!

**Comment by jtpio** [comment]
> I think some of the entries in `doc_snippets` also need updating (e.g. multiple_selection_params).

For the `_MappingSelection`, don't we want to keep support for Mapping / dict? At least for now, and potentially drop it in a separate PR.

**Comment by pbugnion** [comment]
> For the _MultipleSelection, don't we want to keep support for Mapping / dict? At least for now, and potentially drop it in a separate PR.

Ah yes, you're completely right, sorry.

Maybe we should introduce a deprecation ahead of 8?

**Comment by pbugnion** [comment]
Actually, since `_MultipleSelection` also relies on `_make_options`, doesn't this PR break support for dict-like objects in widgets that inherit from `_MultipleSelection`?

**Comment by pbugnion** [comment]
Somewhat unexpectedly, for a dictionary `d = { "a": 5, "b": 8}`, `tuple(d)` returns the dictionary keys:

```
>>> d = {"a": 5, "b": 6}
>>> tuple(d)
#=> ("a", "b")
```

This means that, for someone who's not paid attention to the deprecation warning and still has a dict, their widget won't crash -- it'll just behave really unexpectedly: the dropdown entries will be the same, but the values will just be ignored. I think this is going to be really surprising.

We can either:
1. not care
2. keep the `if isinstance(x, Mapping):` in `_make_options` but raise a `TypeError` if it's x is a `Mapping`.

**Comment by jasongrout** [comment]
> keep the `if isinstance(x, Mapping):` in `_make_options` but raise a `TypeError` if it's x is a `Mapping`.

I'm fine with this, say for one more version.

**Comment by jtpio** [comment]
Thanks @pbugnion for investigating this further.

> Actually, since `_MultipleSelection` also relies on `_make_options`, doesn't this PR break support for dict-like objects in widgets that inherit from `_MultipleSelection`?

It looks like it does yes. We should be able to catch that in the tests though. 

In the long run we'll probably want to drop support for mutable types for `MultipleSelection` widgets too. But it would be better to introduce a `DeprecationWarning` first.

> 2. keep the if isinstance(x, Mapping): in _make_options but raise a TypeError if x is a Mapping.

This could be a good compromise in the meantime.

**Comment by jasongrout** [comment]
Just to be clear here - we are talking about dropping support for [*mapping* types](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping), not mutable types.

**Comment by jasongrout** [comment]
@jtpio - what is the current status of this PR?

**Comment by jtpio** [comment]
It was changed back to a "WIP" to check the points raised by @pbugnion above and see if they can be covered in the tests too:

> Actually, since _MultipleSelection also relies on _make_options, doesn't this PR break support for dict-like objects in widgets that inherit from _MultipleSelection?

**Comment by jtpio** [comment]
We can indeed see that with a `SelectMultiple` widget:

```python
from ipywidgets import SelectMultiple

options = {'a': 'Apples', 'o': 'Oranges', 'p': 'Pears'}

select = SelectMultiple(
    options=options,
    description='Fruits'
)
select
```

Which gives back the list of keys as values: 

image

We might then want to drop support for `Mapping` types as `_MultipleSelection`options too. The `DeprecationWarning` added in #2130 should also be triggered when using a `SelectMultiple` widget in 7.5:

image

**Comment by jtpio** [comment]
> keep the if isinstance(x, Mapping): in _make_options but raise a TypeError if x is a Mapping.

This would at least make the change more visible instead of silently converting the `Mapping` to a `tuple`.

**Comment by jtpio** [comment]
> Currently, the `options` traitlet has type `Any()`. I wonder if we can change it to a `traitlets.List` or some such?

We can also follow this approach and be explicit about the types being accepted. For example with `Union([List(), Tuple(), Set()])`.

**Comment by jtpio** [comment]
The last commit implements the ideas mentioned above:

- raise a `TypeError` for `Mapping` types, since the `DeprecationWarning` was already given on other selection widgets
- this prevents implicit casts to a `tuple`
- `numpy` arrays are still supported

**Comment by jasongrout** [comment]
@jtpio - thanks! @pbugnion - do you have time to review the current changes?

**Comment by pbugnion** [comment]
Great, thank you!

I'll review this in the next couple of days.

**Comment by pbugnion** [comment]
I added a couple of tests to verify the behaviour when setting `.options`, clarified a docstring, and added a note on changes in ipywidgets 8.

RTD is failing because of #2768, but the rest looks good. I'm going to merge this.

Thank you!

**Comment by jtpio** [comment]
Thanks @pbugnion!

**Pull request merged**